packages/flutter/lib/src/cupertino/text_selection_toolbar.dart

```diff
@@ -0,0 +1,1028 @@
+    required this.anchorAbove,
+    required this.anchorBelow,
+    required this.children,
+    this.toolbarBuilder = _defaultToolbarBuilder,
+  }) : assert(children.length > 0),
+       super(key: key);
+
+  /// {@macro flutter.material.TextSelectionToolbar.anchorAbove}
+  final Offset anchorAbove;
+
+  /// {@macro flutter.material.TextSelectionToolbar.anchorBelow}
+  final Offset anchorBelow;
+
+  /// {@macro flutter.material.TextSelectionToolbar.children}
+  ///
+  /// See also:
+  ///   * [CupertinoTextSelectionToolbarButton], which builds a default
+  ///     Cupertino-style text selection toolbar text button.
+  final List<Widget> children;
+
+  /// {@macro flutter.material.TextSelectionToolbar.toolbarBuilder}
+  ///
+  /// The given anchor and isAbove can be used to position an arrow, as in the
+  /// default Cupertino toolbar.
+  final CupertinoToolbarBuilder toolbarBuilder;
+
+  // Builds a toolbar just like the default iOS toolbar, with the right color
+  // background and a rounded cutout with an arrow.
+  static Widget _defaultToolbarBuilder(BuildContext context, Offset anchor, bool isAbove, Widget child) {
+    return _CupertinoTextSelectionToolbarShape(
+      anchor: anchor,
+      isAbove: isAbove,
+      child: DecoratedBox(
+        decoration: const BoxDecoration(color: _kToolbarDividerColor),
+        child: child,
+      ),
+    );
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    assert(debugCheckHasMediaQuery(context));
+    final MediaQueryData mediaQuery = MediaQuery.of(context);
+
+    final double paddingAbove = mediaQuery.padding.top + _kToolbarScreenPadding;
+    final double toolbarHeightNeeded = paddingAbove
+        + _kToolbarContentDistance
+        + _kToolbarHeight;
+    final bool fitsAbove = anchorAbove.dy >= toolbarHeightNeeded;
+
+    const Offset contentPaddingAdjustment = Offset(0.0, _kToolbarContentDistance);
+    final Offset localAdjustment = Offset(_kToolbarScreenPadding, paddingAbove);
+
+    return Padding(
+      padding: EdgeInsets.fromLTRB(
+        _kToolbarScreenPadding,
+        paddingAbove,
+        _kToolbarScreenPadding,
+        _kToolbarScreenPadding,
+      ),
+      child: CustomSingleChildLayout(
+        delegate: TextSelectionToolbarLayoutDelegate(
+          anchorAbove: anchorAbove - localAdjustment - contentPaddingAdjustment,
+          anchorBelow: anchorBelow - localAdjustment + contentPaddingAdjustment,
+        ),
+        child: _CupertinoTextSelectionToolbarContent(
+          anchor: fitsAbove ? anchorAbove : anchorBelow,
+          isAbove: fitsAbove,
+          toolbarBuilder: toolbarBuilder,
+          children: children,
+        ),
+      ),
+    );
+  }
+}
+
+// Clips the child so that it has the shape of the default iOS text selection
+// toolbar, with rounded corners and an arrow pointing at the anchor.
+//
+// The anchor should be in global coordinates.
+class _CupertinoTextSelectionToolbarShape extends SingleChildRenderObjectWidget {
```

Comment:
Most of the rest of the code in this file was copied over from cupertino/text_selection.dart with a little refactoring.  This class in particular was separated out from the positioning logic.